// angular2/src/render/api.js
'use strict';var lang_1 = require('angular2/src/facade/lang');
var collection_1 = require('angular2/src/facade/collection');
/**
 * General notes:
 *
 * The methods for creating / destroying views in this API are used in the AppViewHydrator
 * and RenderViewHydrator as well.
 *
 * We are already parsing expressions on the render side:
 * - this makes the ElementBinders more compact
 *   (e.g. no need to distinguish interpolations from regular expressions from literals)
 * - allows to retrieve which properties should be accessed from the event
 *   by looking at the expression
 * - we need the parse at least for the `template` attribute to match
 *   directives in it
 * - render compiler is not on the critical path as
 *   its output will be stored in precompiled templates.
 */
var EventBinding = (function () {
    function EventBinding(fullName, source) {
        this.fullName = fullName;
        this.source = source;
    }
    return EventBinding;
})();
exports.EventBinding = EventBinding;
(function (PropertyBindingType) {
    PropertyBindingType[PropertyBindingType["PROPERTY"] = 0] = "PROPERTY";
    PropertyBindingType[PropertyBindingType["ATTRIBUTE"] = 1] = "ATTRIBUTE";
    PropertyBindingType[PropertyBindingType["CLASS"] = 2] = "CLASS";
    PropertyBindingType[PropertyBindingType["STYLE"] = 3] = "STYLE";
})(exports.PropertyBindingType || (exports.PropertyBindingType = {}));
var PropertyBindingType = exports.PropertyBindingType;
var ElementPropertyBinding = (function () {
    function ElementPropertyBinding(type, astWithSource, property, unit) {
        if (unit === void 0) { unit = null; }
        this.type = type;
        this.astWithSource = astWithSource;
        this.property = property;
        this.unit = unit;
    }
    return ElementPropertyBinding;
})();
exports.ElementPropertyBinding = ElementPropertyBinding;
var RenderElementBinder = (function () {
    function RenderElementBinder(_a) {
        var _b = _a === void 0 ? {} : _a, index = _b.index, parentIndex = _b.parentIndex, distanceToParent = _b.distanceToParent, directives = _b.directives, nestedProtoView = _b.nestedProtoView, propertyBindings = _b.propertyBindings, variableBindings = _b.variableBindings, eventBindings = _b.eventBindings, readAttributes = _b.readAttributes;
        this.index = index;
        this.parentIndex = parentIndex;
        this.distanceToParent = distanceToParent;
        this.directives = directives;
        this.nestedProtoView = nestedProtoView;
        this.propertyBindings = propertyBindings;
        this.variableBindings = variableBindings;
        this.eventBindings = eventBindings;
        this.readAttributes = readAttributes;
    }
    return RenderElementBinder;
})();
exports.RenderElementBinder = RenderElementBinder;
var DirectiveBinder = (function () {
    function DirectiveBinder(_a) {
        var directiveIndex = _a.directiveIndex, propertyBindings = _a.propertyBindings, eventBindings = _a.eventBindings, hostPropertyBindings = _a.hostPropertyBindings;
        this.directiveIndex = directiveIndex;
        this.propertyBindings = propertyBindings;
        this.eventBindings = eventBindings;
        this.hostPropertyBindings = hostPropertyBindings;
    }
    return DirectiveBinder;
})();
exports.DirectiveBinder = DirectiveBinder;
(function (ViewType) {
    // A view that contains the host element with bound component directive.
    // Contains a COMPONENT view
    ViewType[ViewType["HOST"] = 0] = "HOST";
    // The view of the component
    // Can contain 0 to n EMBEDDED views
    ViewType[ViewType["COMPONENT"] = 1] = "COMPONENT";
    // A view that is embedded into another View via a <template> element
    // inside of a COMPONENT view
    ViewType[ViewType["EMBEDDED"] = 2] = "EMBEDDED";
})(exports.ViewType || (exports.ViewType = {}));
var ViewType = exports.ViewType;
var ProtoViewDto = (function () {
    function ProtoViewDto(_a) {
        var render = _a.render, elementBinders = _a.elementBinders, variableBindings = _a.variableBindings, type = _a.type, textBindings = _a.textBindings, transitiveNgContentCount = _a.transitiveNgContentCount;
        this.render = render;
        this.elementBinders = elementBinders;
        this.variableBindings = variableBindings;
        this.type = type;
        this.textBindings = textBindings;
        this.transitiveNgContentCount = transitiveNgContentCount;
    }
    return ProtoViewDto;
})();
exports.ProtoViewDto = ProtoViewDto;
var RenderDirectiveMetadata = (function () {
    function RenderDirectiveMetadata(_a) {
        var id = _a.id, selector = _a.selector, compileChildren = _a.compileChildren, events = _a.events, hostListeners = _a.hostListeners, hostProperties = _a.hostProperties, hostAttributes = _a.hostAttributes, hostActions = _a.hostActions, properties = _a.properties, readAttributes = _a.readAttributes, type = _a.type, callOnDestroy = _a.callOnDestroy, callOnChange = _a.callOnChange, callOnCheck = _a.callOnCheck, callOnInit = _a.callOnInit, callOnAllChangesDone = _a.callOnAllChangesDone, changeDetection = _a.changeDetection, exportAs = _a.exportAs;
        this.id = id;
        this.selector = selector;
        this.compileChildren = lang_1.isPresent(compileChildren) ? compileChildren : true;
        this.events = events;
        this.hostListeners = hostListeners;
        this.hostAttributes = hostAttributes;
        this.hostProperties = hostProperties;
        this.hostActions = hostActions;
        this.properties = properties;
        this.readAttributes = readAttributes;
        this.type = type;
        this.callOnDestroy = callOnDestroy;
        this.callOnChange = callOnChange;
        this.callOnCheck = callOnCheck;
        this.callOnInit = callOnInit;
        this.callOnAllChangesDone = callOnAllChangesDone;
        this.changeDetection = changeDetection;
        this.exportAs = exportAs;
    }
    Object.defineProperty(RenderDirectiveMetadata, "DIRECTIVE_TYPE", {
        get: function () { return 0; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(RenderDirectiveMetadata, "COMPONENT_TYPE", {
        get: function () { return 1; },
        enumerable: true,
        configurable: true
    });
    RenderDirectiveMetadata.create = function (_a) {
        var id = _a.id, selector = _a.selector, compileChildren = _a.compileChildren, events = _a.events, host = _a.host, properties = _a.properties, readAttributes = _a.readAttributes, type = _a.type, callOnDestroy = _a.callOnDestroy, callOnChange = _a.callOnChange, callOnCheck = _a.callOnCheck, callOnInit = _a.callOnInit, callOnAllChangesDone = _a.callOnAllChangesDone, changeDetection = _a.changeDetection, exportAs = _a.exportAs;
        var hostListeners = new collection_1.Map();
        var hostProperties = new collection_1.Map();
        var hostAttributes = new collection_1.Map();
        var hostActions = new collection_1.Map();
        if (lang_1.isPresent(host)) {
            collection_1.MapWrapper.forEach(host, function (value, key) {
                var matches = lang_1.RegExpWrapper.firstMatch(RenderDirectiveMetadata._hostRegExp, key);
                if (lang_1.isBlank(matches)) {
                    hostAttributes.set(key, value);
                }
                else if (lang_1.isPresent(matches[1])) {
                    hostProperties.set(matches[1], value);
                }
                else if (lang_1.isPresent(matches[2])) {
                    hostListeners.set(matches[2], value);
                }
                else if (lang_1.isPresent(matches[3])) {
                    hostActions.set(matches[3], value);
                }
            });
        }
        return new RenderDirectiveMetadata({
            id: id,
            selector: selector,
            compileChildren: compileChildren,
            events: events,
            hostListeners: hostListeners,
            hostProperties: hostProperties,
            hostAttributes: hostAttributes,
            hostActions: hostActions,
            properties: properties,
            readAttributes: readAttributes,
            type: type,
            callOnDestroy: callOnDestroy,
            callOnChange: callOnChange,
            callOnCheck: callOnCheck,
            callOnInit: callOnInit,
            callOnAllChangesDone: callOnAllChangesDone,
            changeDetection: changeDetection,
            exportAs: exportAs
        });
    };
    // group 1: "property" from "[property]"
    // group 2: "event" from "(event)"
    // group 3: "action" from "@action"
    RenderDirectiveMetadata._hostRegExp = /^(?:(?:\[([^\]]+)\])|(?:\(([^\)]+)\))|(?:@(.+)))$/g;
    return RenderDirectiveMetadata;
})();
exports.RenderDirectiveMetadata = RenderDirectiveMetadata;
// An opaque reference to a render proto ivew
var RenderProtoViewRef = (function () {
    function RenderProtoViewRef() {
    }
    return RenderProtoViewRef;
})();
exports.RenderProtoViewRef = RenderProtoViewRef;
// An opaque reference to a part of a view
var RenderFragmentRef = (function () {
    function RenderFragmentRef() {
    }
    return RenderFragmentRef;
})();
exports.RenderFragmentRef = RenderFragmentRef;
// An opaque reference to a view
var RenderViewRef = (function () {
    function RenderViewRef() {
    }
    return RenderViewRef;
})();
exports.RenderViewRef = RenderViewRef;
/**
 * How the template and styles of a view should be encapsulated.
 */
(function (ViewEncapsulation) {
    /**
     * Emulate scoping of styles by preprocessing the style rules
     * and adding additional attributes to elements. This is the default.
     */
    ViewEncapsulation[ViewEncapsulation["EMULATED"] = 0] = "EMULATED";
    /**
     * Uses the native mechanism of the renderer. For the DOM this means creating a ShadowRoot.
     */
    ViewEncapsulation[ViewEncapsulation["NATIVE"] = 1] = "NATIVE";
    /**
     * Don't scope the template nor the styles.
     */
    ViewEncapsulation[ViewEncapsulation["NONE"] = 2] = "NONE";
})(exports.ViewEncapsulation || (exports.ViewEncapsulation = {}));
var ViewEncapsulation = exports.ViewEncapsulation;
var ViewDefinition = (function () {
    function ViewDefinition(_a) {
        var _b = _a === void 0 ? {} : _a, componentId = _b.componentId, templateAbsUrl = _b.templateAbsUrl, template = _b.template, styleAbsUrls = _b.styleAbsUrls, styles = _b.styles, directives = _b.directives, encapsulation = _b.encapsulation;
        this.componentId = componentId;
        this.templateAbsUrl = templateAbsUrl;
        this.template = template;
        this.styleAbsUrls = styleAbsUrls;
        this.styles = styles;
        this.directives = directives;
        this.encapsulation = lang_1.isPresent(encapsulation) ? encapsulation : ViewEncapsulation.EMULATED;
    }
    return ViewDefinition;
})();
exports.ViewDefinition = ViewDefinition;
var RenderProtoViewMergeMapping = (function () {
    function RenderProtoViewMergeMapping(mergedProtoViewRef, 
        // Number of fragments in the merged ProtoView.
        // Fragments are stored in depth first order of nested ProtoViews.
        fragmentCount, 
        // Mapping from app element index to render element index.
        // Mappings of nested ProtoViews are in depth first order, with all
        // indices for one ProtoView in a consecuitve block.
        mappedElementIndices, 
        // Number of bound render element.
        // Note: This could be more than the original ones
        // as we might have bound a new element for projecting bound text nodes.
        mappedElementCount, 
        // Mapping from app text index to render text index.
        // Mappings of nested ProtoViews are in depth first order, with all
        // indices for one ProtoView in a consecuitve block.
        mappedTextIndices, 
        // Mapping from view index to app element index
        hostElementIndicesByViewIndex, 
        // Number of contained views by view index
        nestedViewCountByViewIndex) {
        this.mergedProtoViewRef = mergedProtoViewRef;
        this.fragmentCount = fragmentCount;
        this.mappedElementIndices = mappedElementIndices;
        this.mappedElementCount = mappedElementCount;
        this.mappedTextIndices = mappedTextIndices;
        this.hostElementIndicesByViewIndex = hostElementIndicesByViewIndex;
        this.nestedViewCountByViewIndex = nestedViewCountByViewIndex;
    }
    return RenderProtoViewMergeMapping;
})();
exports.RenderProtoViewMergeMapping = RenderProtoViewMergeMapping;
var RenderCompiler = (function () {
    function RenderCompiler() {
    }
    /**
     * Creats a ProtoViewDto that contains a single nested component with the given componentId.
     */
    RenderCompiler.prototype.compileHost = function (directiveMetadata) { return null; };
    /**
     * Compiles a single DomProtoView. Non recursive so that
     * we don't need to serialize all possible components over the wire,
     * but only the needed ones based on previous calls.
     */
    RenderCompiler.prototype.compile = function (view) { return null; };
    /**
     * Merges ProtoViews.
     * The first entry of the array is the protoview into which all the other entries of the array
     * should be merged.
     * If the array contains other arrays, they will be merged before processing the parent array.
     * The array must contain an entry for every component and embedded ProtoView of the first entry.
     * @param protoViewRefs List of ProtoViewRefs or nested
     * @return the merge result
     */
    RenderCompiler.prototype.mergeProtoViewsRecursively = function (protoViewRefs) {
        return null;
    };
    return RenderCompiler;
})();
exports.RenderCompiler = RenderCompiler;
var RenderViewWithFragments = (function () {
    function RenderViewWithFragments(viewRef, fragmentRefs) {
        this.viewRef = viewRef;
        this.fragmentRefs = fragmentRefs;
    }
    return RenderViewWithFragments;
})();
exports.RenderViewWithFragments = RenderViewWithFragments;
var Renderer = (function () {
    function Renderer() {
    }
    /**
     * Creates a root host view that includes the given element.
     * Note that the fragmentCount needs to be passed in so that we can create a result
     * synchronously even when dealing with webworkers!
     *
     * @param {RenderProtoViewRef} hostProtoViewRef a RenderProtoViewRef of type
     * ProtoViewDto.HOST_VIEW_TYPE
     * @param {any} hostElementSelector css selector for the host element (will be queried against the
     * main document)
     * @return {RenderViewWithFragments} the created view including fragments
     */
    Renderer.prototype.createRootHostView = function (hostProtoViewRef, fragmentCount, hostElementSelector) {
        return null;
    };
    /**
     * Creates a regular view out of the given ProtoView.
     * Note that the fragmentCount needs to be passed in so that we can create a result
     * synchronously even when dealing with webworkers!
     */
    Renderer.prototype.createView = function (protoViewRef, fragmentCount) {
        return null;
    };
    /**
     * Destroys the given view after it has been dehydrated and detached
     */
    Renderer.prototype.destroyView = function (viewRef) { };
    /**
     * Attaches a fragment after another fragment.
     */
    Renderer.prototype.attachFragmentAfterFragment = function (previousFragmentRef, fragmentRef) { };
    /**
     * Attaches a fragment after an element.
     */
    Renderer.prototype.attachFragmentAfterElement = function (elementRef, fragmentRef) { };
    /**
     * Detaches a fragment.
     */
    Renderer.prototype.detachFragment = function (fragmentRef) { };
    /**
     * Hydrates a view after it has been attached. Hydration/dehydration is used for reusing views
     * inside of the view pool.
     */
    Renderer.prototype.hydrateView = function (viewRef) { };
    /**
     * Dehydrates a view after it has been attached. Hydration/dehydration is used for reusing views
     * inside of the view pool.
     */
    Renderer.prototype.dehydrateView = function (viewRef) { };
    /**
     * Returns the native element at the given location.
     * Attention: In a WebWorker scenario, this should always return null!
     */
    Renderer.prototype.getNativeElementSync = function (location) { return null; };
    /**
     * Sets a property on an element.
     */
    Renderer.prototype.setElementProperty = function (location, propertyName, propertyValue) { };
    /**
     * Sets an attribute on an element.
     */
    Renderer.prototype.setElementAttribute = function (location, attributeName, attributeValue) { };
    /**
     * Sets a class on an element.
     */
    Renderer.prototype.setElementClass = function (location, className, isAdd) { };
    /**
     * Sets a style on an element.
     */
    Renderer.prototype.setElementStyle = function (location, styleName, styleValue) { };
    /**
     * Calls a method on an element.
     */
    Renderer.prototype.invokeElementMethod = function (location, methodName, args) { };
    /**
     * Sets the value of a text node.
     */
    Renderer.prototype.setText = function (viewRef, textNodeIndex, text) { };
    /**
     * Sets the dispatcher for all events of the given view
     */
    Renderer.prototype.setEventDispatcher = function (viewRef, dispatcher) { };
    return Renderer;
})();
exports.Renderer = Renderer;
