// angular2/src/core/application_common.js
'use strict';var di_1 = require('angular2/di');
var lang_1 = require('angular2/src/facade/lang');
var browser_adapter_1 = require('angular2/src/dom/browser_adapter');
var dom_adapter_1 = require('angular2/src/dom/dom_adapter');
var compiler_1 = require('./compiler/compiler');
var reflection_1 = require('angular2/src/reflection/reflection');
var change_detection_1 = require('angular2/src/change_detection/change_detection');
var pipes_1 = require('angular2/pipes');
var exception_handler_1 = require('./exception_handler');
var view_loader_1 = require('angular2/src/render/dom/compiler/view_loader');
var style_url_resolver_1 = require('angular2/src/render/dom/compiler/style_url_resolver');
var style_inliner_1 = require('angular2/src/render/dom/compiler/style_inliner');
var view_resolver_1 = require('./compiler/view_resolver');
var directive_resolver_1 = require('./compiler/directive_resolver');
var pipe_resolver_1 = require('./compiler/pipe_resolver');
var collection_1 = require('angular2/src/facade/collection');
var async_1 = require('angular2/src/facade/async');
var ng_zone_1 = require('angular2/src/core/zone/ng_zone');
var life_cycle_1 = require('angular2/src/core/life_cycle/life_cycle');
var xhr_1 = require('angular2/src/render/xhr');
var xhr_impl_1 = require('angular2/src/render/xhr_impl');
var event_manager_1 = require('angular2/src/render/dom/events/event_manager');
var key_events_1 = require('angular2/src/render/dom/events/key_events');
var hammer_gestures_1 = require('angular2/src/render/dom/events/hammer_gestures');
var component_url_mapper_1 = require('angular2/src/core/compiler/component_url_mapper');
var url_resolver_1 = require('angular2/src/services/url_resolver');
var app_root_url_1 = require('angular2/src/services/app_root_url');
var anchor_based_app_root_url_1 = require('angular2/src/services/anchor_based_app_root_url');
var dynamic_component_loader_1 = require('angular2/src/core/compiler/dynamic_component_loader');
var testability_1 = require('angular2/src/core/testability/testability');
var view_pool_1 = require('angular2/src/core/compiler/view_pool');
var view_manager_1 = require('angular2/src/core/compiler/view_manager');
var view_manager_utils_1 = require('angular2/src/core/compiler/view_manager_utils');
var view_listener_1 = require('angular2/src/core/compiler/view_listener');
var proto_view_factory_1 = require('angular2/src/core/compiler/proto_view_factory');
var api_1 = require('angular2/src/render/api');
var render_1 = require('angular2/src/render/render');
var element_schema_registry_1 = require('angular2/src/render/dom/schema/element_schema_registry');
var dom_element_schema_registry_1 = require('angular2/src/render/dom/schema/dom_element_schema_registry');
var shared_styles_host_1 = require('angular2/src/render/dom/view/shared_styles_host');
var view_ref_1 = require('angular2/src/core/compiler/view_ref');
var application_tokens_1 = require('./application_tokens');
var wtf_init_1 = require('../profile/wtf_init');
var platform_bindings_1 = require('./platform_bindings');
var _rootInjector;
// Contains everything that is safe to share between applications.
var _rootBindings = [di_1.bind(reflection_1.Reflector).toValue(reflection_1.reflector), testability_1.TestabilityRegistry];
function _injectorBindings(appComponentType) {
    var bestChangeDetection = change_detection_1.DynamicChangeDetection;
    if (change_detection_1.PreGeneratedChangeDetection.isSupported()) {
        bestChangeDetection = change_detection_1.PreGeneratedChangeDetection;
    }
    else if (change_detection_1.JitChangeDetection.isSupported()) {
        bestChangeDetection = change_detection_1.JitChangeDetection;
    }
    return [
        di_1.bind(render_1.DOCUMENT)
            .toValue(dom_adapter_1.DOM.defaultDoc()),
        di_1.bind(render_1.DOM_REFLECT_PROPERTIES_AS_ATTRIBUTES).toValue(false),
        di_1.bind(application_tokens_1.APP_COMPONENT).toValue(appComponentType),
        di_1.bind(application_tokens_1.APP_COMPONENT_REF_PROMISE)
            .toFactory(function (dynamicComponentLoader, injector, testability, registry) {
            // TODO(rado): investigate whether to support bindings on root component.
            return dynamicComponentLoader.loadAsRoot(appComponentType, null, injector)
                .then(function (componentRef) {
                registry.registerApplication(componentRef.location.nativeElement, testability);
                return componentRef;
            });
        }, [dynamic_component_loader_1.DynamicComponentLoader, di_1.Injector, testability_1.Testability, testability_1.TestabilityRegistry]),
        di_1.bind(appComponentType)
            .toFactory(function (p) { return p.then(function (ref) { return ref.instance; }); }, [application_tokens_1.APP_COMPONENT_REF_PROMISE]),
        di_1.bind(life_cycle_1.LifeCycle).toFactory(function (exceptionHandler) { return new life_cycle_1.LifeCycle(null, lang_1.assertionsEnabled()); }, [exception_handler_1.ExceptionHandler]),
        di_1.bind(event_manager_1.EventManager)
            .toFactory(function (ngZone) {
            var plugins = [new hammer_gestures_1.HammerGesturesPlugin(), new key_events_1.KeyEventsPlugin(), new event_manager_1.DomEventsPlugin()];
            return new event_manager_1.EventManager(plugins, ngZone);
        }, [ng_zone_1.NgZone]),
        render_1.DomRenderer,
        di_1.bind(api_1.Renderer).toAlias(render_1.DomRenderer),
        render_1.APP_ID_RANDOM_BINDING,
        render_1.TemplateCloner,
        di_1.bind(render_1.MAX_IN_MEMORY_ELEMENTS_PER_TEMPLATE).toValue(20),
        render_1.DefaultDomCompiler,
        di_1.bind(element_schema_registry_1.ElementSchemaRegistry).toValue(new dom_element_schema_registry_1.DomElementSchemaRegistry()),
        di_1.bind(api_1.RenderCompiler).toAlias(render_1.DefaultDomCompiler),
        shared_styles_host_1.DomSharedStylesHost,
        di_1.bind(shared_styles_host_1.SharedStylesHost).toAlias(shared_styles_host_1.DomSharedStylesHost),
        proto_view_factory_1.ProtoViewFactory,
        view_pool_1.AppViewPool,
        di_1.bind(view_pool_1.APP_VIEW_POOL_CAPACITY).toValue(10000),
        view_manager_1.AppViewManager,
        view_manager_utils_1.AppViewManagerUtils,
        view_listener_1.AppViewListener,
        compiler_1.Compiler,
        compiler_1.CompilerCache,
        view_resolver_1.ViewResolver,
        pipes_1.DEFAULT_PIPES,
        di_1.bind(change_detection_1.IterableDiffers).toValue(change_detection_1.defaultIterableDiffers),
        di_1.bind(change_detection_1.KeyValueDiffers).toValue(change_detection_1.defaultKeyValueDiffers),
        di_1.bind(change_detection_1.ChangeDetection).toClass(bestChangeDetection),
        view_loader_1.ViewLoader,
        directive_resolver_1.DirectiveResolver,
        pipe_resolver_1.PipeResolver,
        change_detection_1.Parser,
        change_detection_1.Lexer,
        platform_bindings_1.EXCEPTION_BINDING,
        di_1.bind(xhr_1.XHR).toValue(new xhr_impl_1.XHRImpl()),
        component_url_mapper_1.ComponentUrlMapper,
        url_resolver_1.UrlResolver,
        style_url_resolver_1.StyleUrlResolver,
        style_inliner_1.StyleInliner,
        dynamic_component_loader_1.DynamicComponentLoader,
        testability_1.Testability,
        anchor_based_app_root_url_1.AnchorBasedAppRootUrl,
        di_1.bind(app_root_url_1.AppRootUrl).toAlias(anchor_based_app_root_url_1.AnchorBasedAppRootUrl)
    ];
}
function createNgZone() {
    return new ng_zone_1.NgZone({ enableLongStackTrace: lang_1.assertionsEnabled() });
}
exports.createNgZone = createNgZone;
/**
 * Bootstrapping for Angular applications.
 *
 * You instantiate an Angular application by explicitly specifying a component to use as the root
 * component for your
 * application via the `bootstrap()` method.
 *
 * ## Simple Example
 *
 * Assuming this `index.html`:
 *
 * ```html
 * <html>
 *   <!-- load Angular script tags here. -->
 *   <body>
 *     <my-app>loading...</my-app>
 *   </body>
 * </html>
 * ```
 *
 * An application is bootstrapped inside an existing browser DOM, typically `index.html`. Unlike
 * Angular 1, Angular 2
 * does not compile/process bindings in `index.html`. This is mainly for security reasons, as well
 * as architectural
 * changes in Angular 2. This means that `index.html` can safely be processed using server-side
 * technologies such as
 * bindings. Bindings can thus use double-curly `{{ syntax }}` without collision from Angular 2
 * component double-curly
 * `{{ syntax }}`.
 *
 * We can use this script code:
 *
 * ```
 * @Component({
 *    selector: 'my-app'
 * })
 * @View({
 *    template: 'Hello {{ name }}!'
 * })
 * class MyApp {
 *   name:string;
 *
 *   constructor() {
 *     this.name = 'World';
 *   }
 * }
 *
 * main() {
 *   return bootstrap(MyApp);
 * }
 * ```
 *
 * When the app developer invokes `bootstrap()` with the root component `MyApp` as its argument,
 * Angular performs the
 * following tasks:
 *
 *  1. It uses the component's `selector` property to locate the DOM element which needs to be
 * upgraded into
 *     the angular component.
 *  2. It creates a new child injector (from the platform injector). Optionally, you can also
 * override the injector configuration for an app by
 * invoking `bootstrap` with the `componentInjectableBindings` argument.
 *  3. It creates a new `Zone` and connects it to the angular application's change detection domain
 * instance.
 *  4. It creates a shadow DOM on the selected component's host element and loads the template into
 * it.
 *  5. It instantiates the specified component.
 *  6. Finally, Angular performs change detection to apply the initial data bindings for the
 * application.
 *
 *
 * ## Instantiating Multiple Applications on a Single Page
 *
 * There are two ways to do this.
 *
 *
 * ### Isolated Applications
 *
 * Angular creates a new application each time that the `bootstrap()` method is invoked. When
 * multiple applications
 * are created for a page, Angular treats each application as independent within an isolated change
 * detection and
 * `Zone` domain. If you need to share data between applications, use the strategy described in the
 * next
 * section, "Applications That Share Change Detection."
 *
 *
 * ### Applications That Share Change Detection
 *
 * If you need to bootstrap multiple applications that share common data, the applications must
 * share a common
 * change detection and zone. To do that, create a meta-component that lists the application
 * components in its template.
 * By only invoking the `bootstrap()` method once, with the meta-component as its argument, you
 * ensure that only a
 * single change detection zone is created and therefore data can be shared across the applications.
 *
 *
 * ## Platform Injector
 *
 * When working within a browser window, there are many singleton resources: cookies, title,
 * location, and others.
 * Angular services that represent these resources must likewise be shared across all Angular
 * applications that
 * occupy the same browser window.  For this reason, Angular creates exactly one global platform
 * injector which stores
 * all shared services, and each angular application injector has the platform injector as its
 * parent.
 *
 * Each application has its own private injector as well. When there are multiple applications on a
 * page, Angular treats
 * each application injector's services as private to that application.
 *
 *
 * # API
 * - `appComponentType`: The root component which should act as the application. This is a reference
 * to a `Type`
 *   which is annotated with `@Component(...)`.
 * - `componentInjectableBindings`: An additional set of bindings that can be added to the app
 * injector
 * to override default injection behavior.
 * - `errorReporter`: `function(exception:any, stackTrace:string)` a default error reporter for
 * unhandled exceptions.
 *
 * Returns a `Promise` of {@link ApplicationRef}.
 */
function commonBootstrap(appComponentType, componentInjectableBindings) {
    if (componentInjectableBindings === void 0) { componentInjectableBindings = null; }
    browser_adapter_1.BrowserDomAdapter.makeCurrent();
    wtf_init_1.wtfInit();
    var bootstrapProcess = async_1.PromiseWrapper.completer();
    var zone = createNgZone();
    zone.run(function () {
        var exceptionHandler;
        try {
            var appInjector = _createAppInjector(appComponentType, componentInjectableBindings, zone);
            exceptionHandler = appInjector.get(exception_handler_1.ExceptionHandler);
            zone.overrideOnErrorHandler(function (e, s) { return exceptionHandler.call(e, s); });
            var compRefToken = appInjector.get(application_tokens_1.APP_COMPONENT_REF_PROMISE);
            var tick = function (componentRef) {
                var appChangeDetector = view_ref_1.internalView(componentRef.hostView).changeDetector;
                // retrieve life cycle: may have already been created if injected in root component
                var lc = appInjector.get(life_cycle_1.LifeCycle);
                lc.registerWith(zone, appChangeDetector);
                lc.tick(); // the first tick that will bootstrap the app
                bootstrapProcess.resolve(new ApplicationRef(componentRef, appComponentType, appInjector));
            };
            var tickResult = async_1.PromiseWrapper.then(compRefToken, tick);
            async_1.PromiseWrapper.then(tickResult, function (_) { }); // required for Dart to trigger the default error handler
            async_1.PromiseWrapper.then(tickResult, null, function (err, stackTrace) { bootstrapProcess.reject(err, stackTrace); });
        }
        catch (e) {
            if (lang_1.isPresent(exceptionHandler)) {
                exceptionHandler.call(e, e.stack);
            }
            else {
                // The error happened during the creation of an injector, most likely because of a bug in
                // DI.
                // We cannot use the provided exception handler, so we default to writing to the DOM.
                dom_adapter_1.DOM.logError(e);
            }
            bootstrapProcess.reject(e, e.stack);
        }
    });
    return bootstrapProcess.promise;
}
exports.commonBootstrap = commonBootstrap;
/**
 * Represents a Angular's representation of an Application.
 *
 * `ApplicationRef` represents a running application instance. Use it to retrieve the host
 * component, injector,
 * or dispose of an application.
 */
var ApplicationRef = (function () {
    /**
     * @private
     */
    function ApplicationRef(hostComponent, hostComponentType, injector) {
        this._hostComponent = hostComponent;
        this._injector = injector;
        this._hostComponentType = hostComponentType;
    }
    Object.defineProperty(ApplicationRef.prototype, "hostComponentType", {
        /**
         * Returns the current {@link ComponentMetadata} type.
         */
        get: function () { return this._hostComponentType; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ApplicationRef.prototype, "hostComponent", {
        /**
         * Returns the current {@link ComponentMetadata} instance.
         */
        get: function () { return this._hostComponent.instance; },
        enumerable: true,
        configurable: true
    });
    /**
     * Dispose (un-load) the application.
     */
    ApplicationRef.prototype.dispose = function () {
        // TODO: We also need to clean up the Zone, ... here!
        this._hostComponent.dispose();
    };
    Object.defineProperty(ApplicationRef.prototype, "injector", {
        /**
         * Returns the root application {@link Injector}.
         */
        get: function () { return this._injector; },
        enumerable: true,
        configurable: true
    });
    return ApplicationRef;
})();
exports.ApplicationRef = ApplicationRef;
function _createAppInjector(appComponentType, bindings, zone) {
    if (lang_1.isBlank(_rootInjector))
        _rootInjector = di_1.Injector.resolveAndCreate(_rootBindings);
    var mergedBindings = lang_1.isPresent(bindings) ? collection_1.ListWrapper.concat(_injectorBindings(appComponentType), bindings) :
        _injectorBindings(appComponentType);
    mergedBindings.push(di_1.bind(ng_zone_1.NgZone).toValue(zone));
    return _rootInjector.resolveAndCreateChild(mergedBindings);
}
