// angular2/src/core/metadata.js
'use strict';/**
 * This indirection is needed to free up Component, etc symbols in the public API
 * to be used by the decorator versions of these annotations.
 */
var di_1 = require('./metadata/di');
exports.QueryMetadata = di_1.QueryMetadata;
exports.ViewQueryMetadata = di_1.ViewQueryMetadata;
exports.AttributeMetadata = di_1.AttributeMetadata;
var directives_1 = require('./metadata/directives');
exports.ComponentMetadata = directives_1.ComponentMetadata;
exports.DirectiveMetadata = directives_1.DirectiveMetadata;
exports.PipeMetadata = directives_1.PipeMetadata;
exports.LifecycleEvent = directives_1.LifecycleEvent;
var view_1 = require('./metadata/view');
exports.ViewMetadata = view_1.ViewMetadata;
exports.ViewEncapsulation = view_1.ViewEncapsulation;
var di_2 = require('./metadata/di');
var directives_2 = require('./metadata/directives');
var view_2 = require('./metadata/view');
var decorators_1 = require('../util/decorators');
/**
 * {@link ComponentMetadata} factory function.
 */
exports.Component = decorators_1.makeDecorator(directives_2.ComponentMetadata, function (fn) { return fn.View = exports.View; });
/**
 * {@link DirectiveMetadata} factory function.
 */
exports.Directive = decorators_1.makeDecorator(directives_2.DirectiveMetadata);
/**
 * {@link ViewMetadata} factory function.
 */
exports.View = decorators_1.makeDecorator(view_2.ViewMetadata, function (fn) { return fn.View = exports.View; });
/**
 * {@link AttributeMetadata} factory function.
 */
exports.Attribute = decorators_1.makeParamDecorator(di_2.AttributeMetadata);
/**
 * {@link QueryMetadata} factory function.
 */
exports.Query = decorators_1.makeParamDecorator(di_2.QueryMetadata);
/**
 * {@link ViewQueryMetadata} factory function.
 */
exports.ViewQuery = decorators_1.makeParamDecorator(di_2.ViewQueryMetadata);
/**
 * {@link PipeMetadata} factory function.
 */
exports.Pipe = decorators_1.makeDecorator(directives_2.PipeMetadata);
