// angular2/pipes.js
'use strict';/**
 * @module
 * @description
 * This module provides advanced support for extending change detection.
 */
var uppercase_pipe_1 = require('./src/pipes/uppercase_pipe');
exports.UpperCasePipe = uppercase_pipe_1.UpperCasePipe;
var lowercase_pipe_1 = require('./src/pipes/lowercase_pipe');
exports.LowerCasePipe = lowercase_pipe_1.LowerCasePipe;
var async_pipe_1 = require('./src/pipes/async_pipe');
exports.AsyncPipe = async_pipe_1.AsyncPipe;
var json_pipe_1 = require('./src/pipes/json_pipe');
exports.JsonPipe = json_pipe_1.JsonPipe;
var date_pipe_1 = require('./src/pipes/date_pipe');
exports.DatePipe = date_pipe_1.DatePipe;
var number_pipe_1 = require('./src/pipes/number_pipe');
exports.DecimalPipe = number_pipe_1.DecimalPipe;
exports.PercentPipe = number_pipe_1.PercentPipe;
exports.CurrencyPipe = number_pipe_1.CurrencyPipe;
var limit_to_pipe_1 = require('./src/pipes/limit_to_pipe');
exports.LimitToPipe = limit_to_pipe_1.LimitToPipe;
var default_pipes_1 = require('./src/pipes/default_pipes');
exports.DEFAULT_PIPES_TOKEN = default_pipes_1.DEFAULT_PIPES_TOKEN;
exports.DEFAULT_PIPES = default_pipes_1.DEFAULT_PIPES;
